decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java

```diff
@@ -0,0 +1,305 @@
+            return graphEdge;
+        }
+        return null;
+    }
+
+    /**
+     * Backtrack the graph to identify where the connection is broken and return an unmarked edge with the end node same as the
+     * start node of the last tracked edge. if all edges are already marked then it will try to find an edge from the source slice.
+     */
+    private GraphEdge getEdgeByBackwardTracking() {
+
+        GraphEdge<T, L> edge = graphEdge;
+        GraphEdge<T, L> parent = null;
+        do {
+            parent = edge.getSourceNode().getIncomingEdge()
+                    .stream()
+                    .filter(incomingEdge -> incomingEdge.getEdgeColor().equals(EntityColor.GREEN)).findFirst().orElse(null);
+            if (parent != null) {
+                edge = parent;
+            }
+            --currentSourceSliceIndex;
+        } while (parent != null);
+
+        if (edge.getSourceNode().getIncomingEdge()
+                .stream().count() == 0) {
+            return null;
+        }
+
+        edge = edge.getSourceNode().getIncomingEdge()
+                .stream()
+                .filter(incomingEdge -> !incomingEdge.getEdgeColor().equals(EntityColor.RED)).findFirst().orElse(null);
+
+
+        if (edge == null) {
+            return getNextEdgeFromNormalFlow();
+        }
+
+        return edge;
+    }
+
+
+    /**
+     * if there are any green successor to the last edge then it will try to find the last green edge in the successor links
+     * then return an unmarked edge starting from the end node of the final green edge, if there are no such unmarked edges then it will try
+     * to find the best candidate from slice which contains the start node of the last edge in the chain
+     */
+    private GraphEdge<T, L> getEdgeByForwardTracking() {
+        GraphEdge<T, L> edge = graphEdge;
+        GraphEdge<T, L> successor = null;
+        do {
+            successor = (edge.getDestinationNode().getOutgoingEdge()
+                    .stream()
+                    .filter(outgoingEdge -> outgoingEdge.getEdgeColor().equals(EntityColor.GREEN)).findFirst().orElse(null));
+            if (successor != null) {
+                edge = successor;
+                lastSuccessfulEdge = edge;
+            }
+            ++currentSourceSliceIndex;
+        } while (successor != null);
+
+
+        if (edge.getDestinationNode().getOutgoingEdge()
+                .stream().count() == 0) {
+            return null;
+        }
+
+        edge = edge.getDestinationNode().getOutgoingEdge()
+                .stream()
+                .filter(outgoingEdge -> !outgoingEdge.getEdgeColor().equals(EntityColor.RED)).findFirst().orElse(null);
+
+        if (edge == null) {
+            return getNextEdgeFromNormalFlow();
+        }
+
+        return edge;
+    }
+
+    /**
+     * retrieve the next best edge based on the users opinion of the last best edge
+     */
+    private GraphEdge<T, L> edgeFormGraph() {
```

Comment:
Should this be called `edgeFromGraph`?

Reply:
31890348b9aef835f9f9065ddd8625de1cfab0ea